src/components/App.tsx

```diff
@@ -2450,7 +2451,28 @@ class App extends React.Component<any, AppState> {
             y - dragOffsetXY[1],
             this.state.gridSize,
           );
-          dragSelectedElements(selectedElements, dragX, dragY);
+
+          if (this.state.draggingElement && event.shiftKey) {
+            if (!selectedElementsOriginalPosition) {
+              selectedElementsOriginalPosition = selectedElements.map((e) => ({
+                x: e.x,
+                y: e.y,
+              }));
+            }
+
+            const dragDistanceX = Math.abs(x - originX);
+            const dragDistanceY = Math.abs(y - originY);
+            const lockX = dragDistanceX < dragDistanceY;
+            const lockY = dragDistanceX > dragDistanceY;
+
+            for (var [i, e] of selectedElements.entries()) {
+              mutateElement(e, selectedElementsOriginalPosition[i]);
+            }
+
```

Comment:
Wow, you did the fix pretty quickly!

Would it possible to put the new code as much as possible in dragElements.ts?
I think App.tsx is too big now and it'd be better to split code into smaller files.

Also, it would be nice to avoid calling `mutateElement` twice. It's more readable and doesn't rely on batched updates.

Reply:
What would you suggest? Since quite a bit of data will still be available in the local scope of App.tsx. I could abstract parts of it into a separate function? Or add more parameters?

At that point I would propose exporting two functions, the current `dragSelectedElements()` and a new `dragSelectedElementsLockDirection()` which additionally takes the `selectedElementsOriginalPosition`, `originX`, `originY`, `x`, and `y` values. In this way we can keep the original free dragging small and fast and can work with strong required params and predictable behavior on the new one. Thoughts?

Reply:
Exporting two functions is totally fine. I would probably export one function, which calls either `dragSelectedElements` or `dragSelectedElementsLocked`, if it were to reduce the lines in App.tsx. 

> I could abstract parts of it into a separate function?

Like `getDragOffsetXY`? Anyway, I'd encourage separating as long as it's well readable.

We don't need to do separation too much. If you could make a little effort for it, it would be great!